File: web/src/stores/cexSpotPerp.ts
import { computed, ref } from 'vue'
import { defineStore } from 'pinia'
import api from '@/api'

export type SpotPerpDirection = 'spot_long_perp_short' | 'spot_short_inventory_perp_long'
export type MarketStatus = 'running' | 'halted'

export interface SpotPerpConfig {
  symbols: string[]
  exchanges: string[]
  exchangeSymbols: Record<string, string[]>
  leverage: number
  maxLeverage: number
  minNetProfitRate: number
  carryFundingIntervals: number
}

export interface SimAccount {
  exchange: string
  usdt: number
  perpUsdt: number
  frozenUsdt: number
  spotBalances: Record<string, number>
  perpPositions: Record<string, number>
}

export interface MarketQuote {
  exchange: string
  symbol: string
  marketType: 'spot' | 'perp'
  bid: number
  ask: number
  last: number
  fundingRate: number
  timestamp: number
  ageMillis: number
  stale: boolean
}

export interface SpotPerpOpportunity {
  id: string
  symbol: string
  direction: SpotPerpDirection
  spotExchange: string
  perpExchange: string
  spotPrice: number
  perpPrice: number
  notional: number
  basisAmount: number
  fundingAmount: number
  feeCost: number
  slippage: number
  safetyBuffer: number
  netProfit: number
  profitRate: number
  carryFundingAmount: number
  carryNetProfit: number
  carryProfitRate: number
  carryFundingIntervals: number
  status: 'ready' | 'watch' | 'blocked'
  blockReason?: string
}

export interface SimPosition {
  id: string
  opportunityId: string
  symbol: string
  direction: SpotPerpDirection
  spotExchange: string
  perpExchange: string
  quantity: number
  notional: number
  margin: number
  spotPrice: number
  perpPrice: number
  netProfit: number
  realizedPnL: number
  openedAt: number
  closedAt?: number
  status: 'open' | 'closing' | 'closed'
}

export interface CloseAction {
  id: string
  positionId: string
  reason: string
  spotAction: string
  perpAction: string
  createdAt: number
}

export interface SpotPerpAutomation {
  enabled: boolean
  autoOpen: boolean
  autoClose: boolean
  openMinProfitRate: number
  closeMinProfitRate: number
  maxHoldSeconds: number
  maxOpenPositions: number
  checkIntervalMillis: number
}

export interface SpotPerpAutoStats {
  autoOpenCount: number
  autoCloseCount: number
  winCount: number
  lossCount: number
  totalProfit: number
  averageProfit: number
  winRate: number
  lastActionAt: number
  lastActionError: string
}

export interface OpportunityLog {
  key: string
  id: string
  symbol: string
  direction: SpotPerpDirection
  spotExchange: string
  perpExchange: string
  firstSeenAt: number
  lastSeenAt: number
  seenCount: number
  bestProfit: number
  bestProfitRate: number
  lastProfit: number
  lastProfitRate: number
  lastStatus: 'ready' | 'watch' | 'blocked'
  lastBlockReason: string
  autoOpenedCount: number
  autoRejectedNote: string
}

export interface AutoTrade {
  id: string
  positionId: string
  opportunity: string
  symbol: string
  direction: SpotPerpDirection
  spotExchange: string
  perpExchange: string
  action: 'open' | 'close'
  reason: string
  quantity: number
  notional: number
  margin: number
  spotValue: number
  capitalUsed: number
  profit: number
  profitRate: number
  createdAt: number
}

const defaultTopSymbols = [
  'BTCUSDT', 'ETHUSDT', 'BNBUSDT', 'SOLUSDT', 'XRPUSDT',
  'DOGEUSDT', 'ADAUSDT', 'TRXUSDT', 'LINKUSDT', 'AVAXUSDT',
  'TONUSDT', 'SHIBUSDT', 'DOTUSDT', 'BCHUSDT', 'LTCUSDT',
  'UNIUSDT', 'NEARUSDT', 'APTUSDT', 'ICPUSDT', 'ETCUSDT',
]

interface SimulationSnapshot {
  status: MarketStatus
  haltReason: string
  lastQuoteAt: number
  marketErrors: Record<string, string>
  wsStatus: Record<string, string>
  wsErrors: Record<string, string>
  config: SpotPerpConfig
  accounts: SimAccount[]
  quotes: MarketQuote[]
  opportunities: SpotPerpOpportunity[]
  positions: SimPosition[]
  closeActions: CloseAction[]
  opportunityLogs: OpportunityLog[]
  autoTrades: AutoTrade[]
  automation: SpotPerpAutomation
  autoStats: SpotPerpAutoStats
  pnl: {
    realizedPnL: number
    unrealizedPnL: number
    totalPnL: number
    openNotional: number
  }
}

export const useCexSpotPerpStore = defineStore('cexSpotPerp', () => {
  const loading = ref(false)
  const wsConnected = ref(false)
  const status = ref<MarketStatus>('running')
  const haltReason = ref('')
  const config = ref<SpotPerpConfig>({
    symbols: [...defaultTopSymbols],
    exchanges: ['binance', 'okx', 'bitget'],
    exchangeSymbols: {
      binance: [...defaultTopSymbols],
      okx: [...defaultTopSymbols],
      bitget: [...defaultTopSymbols],
    },
    leverage: 3,
    maxLeverage: 3,
    minNetProfitRate: 0.05,
    carryFundingIntervals: 6,
  })
  const accounts = ref<SimAccount[]>([])
  const quotes = ref<MarketQuote[]>([])
  const opportunities = ref<SpotPerpOpportunity[]>([])
  const positions = ref<SimPosition[]>([])
  const closeActions = ref<CloseAction[]>([])
  const opportunityLogs = ref<OpportunityLog[]>([])
  const autoTrades = ref<AutoTrade[]>([])
  const automation = ref<SpotPerpAutomation>({
    enabled: false,
    autoOpen: true,
    autoClose: true,
    openMinProfitRate: 0.05,
    closeMinProfitRate: 0,
    maxHoldSeconds: 300,
    maxOpenPositions: 3,
    checkIntervalMillis: 1000,
  })
  const autoStats = ref<SpotPerpAutoStats>({
    autoOpenCount: 0,
    autoCloseCount: 0,
    winCount: 0,
    lossCount: 0,
    totalProfit: 0,
    averageProfit: 0,
    winRate: 0,
    lastActionAt: 0,
    lastActionError: '',
  })
  const lastQuoteAt = ref(0)
  const marketErrors = ref<Record<string, string>>({})
  const wsStatus = ref<Record<string, string>>({})
  const wsErrors = ref<Record<string, string>>({})
  const pnl = ref({
    realizedPnL: 0,
    unrealizedPnL: 0,
    totalPnL: 0,
    openNotional: 0,
  })

  const totalSpotUsdt = computed(() => accounts.value.reduce((sum, account) => sum + account.usdt, 0))
  const totalPerpUsdt = computed(() => accounts.value.reduce((sum, account) => sum + account.perpUsdt, 0))
  const totalFrozenUsdt = computed(() => accounts.value.reduce((sum, account) => sum + account.frozenUsdt, 0))
  const readyOpportunities = computed(() => opportunities.value.filter(opp => opp.status === 'ready'))
  const watchOpportunities = computed(() => opportunities.value.filter(opp => opp.status === 'watch'))
  const openPositions = computed(() => positions.value.filter(position => position.status === 'open'))
  let ws: WebSocket | null = null
  let reconnectTimer: number | undefined

  function hydrate(snapshot: SimulationSnapshot) {
    status.value = snapshot.status
    haltReason.value = snapshot.haltReason || ''
    config.value = snapshot.config || config.value
    accounts.value = snapshot.accounts || []
    quotes.value = snapshot.quotes || []
    opportunities.value = snapshot.opportunities || []
    positions.value = snapshot.positions || []
    closeActions.value = snapshot.closeActions || []
    opportunityLogs.value = snapshot.opportunityLogs || []
    autoTrades.value = snapshot.autoTrades || []
    automation.value = snapshot.automation || automation.value
    autoStats.value = snapshot.autoStats || autoStats.value
    lastQuoteAt.value = snapshot.lastQuoteAt || 0
    marketErrors.value = snapshot.marketErrors || {}
    wsStatus.value = snapshot.wsStatus || {}
    wsErrors.value = snapshot.wsErrors || {}
    pnl.value = snapshot.pnl || { realizedPnL: 0, unrealizedPnL: 0, totalPnL: 0, openNotional: 0 }
  }

  async function fetchSimulation() {
    loading.value = true
    try {
      const response = await api.get('/cex-spot-perp') as unknown as SimulationSnapshot
      hydrate(response)
    } finally {
      loading.value = false
    }
  }

  function connectWebSocket() {
    if (ws && (ws.readyState === WebSocket.OPEN || ws.readyState === WebSocket.CONNECTING)) return
    if (reconnectTimer) {
      window.clearTimeout(reconnectTimer)
      reconnectTimer = undefined
    }

    const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:'
    ws = new WebSocket(`${protocol}//${window.location.host}/api/v1/cex-spot-perp/ws`)

    ws.onopen = () => {
      wsConnected.value = true
    }
    ws.onmessage = (event) => {
      const message = JSON.parse(event.data) as { type: string; data: SimulationSnapshot }
      if (message.type === 'cex_spot_perp_snapshot') {
        hydrate(message.data)
      }
    }
    ws.onclose = () => {
      wsConnected.value = false
      ws = null
      reconnectTimer = window.setTimeout(connectWebSocket, 2000)
    }
    ws.onerror = () => {
      ws?.close()
    }
  }

  function disconnectWebSocket() {
    if (reconnectTimer) {
      window.clearTimeout(reconnectTimer)
      reconnectTimer = undefined
    }
    wsConnected.value = false
    ws?.close()
    ws = null
  }

  async function executeOpportunity(opp: SpotPerpOpportunity) {
    if (status.value === 'halted' || opp.status !== 'ready') return
    const response = await api.post(`/cex-spot-perp/opportunities/${encodeURIComponent(opp.id)}/execute`) as unknown as SimulationSnapshot
    hydrate(response)
  }

  async function closePosition(positionId: string, reason = '手动平仓') {
    const response = await api.post(`/cex-spot-perp/positions/${encodeURIComponent(positionId)}/close`, { reason }) as unknown as SimulationSnapshot
    hydrate(response)
  }

  async function triggerCircuitBreaker(reason: string) {
    const response = await api.post('/cex-spot-perp/circuit-breaker', { reason }) as unknown as SimulationSnapshot
    hydrate(response)
  }

  async function resumeSimulation() {
    const response = await api.post('/cex-spot-perp/resume') as unknown as SimulationSnapshot
    hydrate(response)
  }

  async function updateConfig(nextConfig: SpotPerpConfig) {
    const response = await api.put('/cex-spot-perp/config', nextConfig) as unknown as SimulationSnapshot
    hydrate(response)
  }

  async function updateAutomation(nextAutomation: SpotPerpAutomation) {
    const response = await api.put('/cex-spot-perp/automation', nextAutomation) as unknown as SimulationSnapshot
    hydrate(response)
  }

  async function updateAccount(account: SimAccount) {
    const response = await api.put(`/cex-spot-perp/accounts/${encodeURIComponent(account.exchange)}`, {
      usdt: account.usdt,
      perpUsdt: account.perpUsdt,
      spotBalances: account.spotBalances,
    }) as unknown as SimulationSnapshot
    hydrate(response)
  }

  async function transferAccountUSDT(exchange: string, from: 'spot' | 'perp', to: 'spot' | 'perp', amount: number) {
    const response = await api.post(`/cex-spot-perp/accounts/${encodeURIComponent(exchange)}/transfer`, {
      from,
      to,
      amount,
    }) as unknown as SimulationSnapshot
    hydrate(response)
  }

  async function resetAccounts() {
    const response = await api.post('/cex-spot-perp/accounts/reset') as unknown as SimulationSnapshot
    hydrate(response)
  }

  return {
    loading,
    wsConnected,
    status,
    haltReason,
    config,
    accounts,
    quotes,
    opportunities,
    positions,
    closeActions,
    opportunityLogs,
    autoTrades,
    automation,
    autoStats,
    lastQuoteAt,
    marketErrors,
    wsStatus,
    wsErrors,
    pnl,
    totalSpotUsdt,
    totalPerpUsdt,
    totalFrozenUsdt,
    readyOpportunities,
    watchOpportunities,
    openPositions,
    fetchSimulation,
    connectWebSocket,
    disconnectWebSocket,
    executeOpportunity,
    closePosition,
    triggerCircuitBreaker,
    resumeSimulation,
    updateConfig,
    updateAutomation,
    updateAccount,
    transferAccountUSDT,
    resetAccounts,
  }
})

File: internal/model/cex_spot_perp.go
package model

import "time"

type CEXSpotPerpAutomationSetting struct {
	Name                string    `gorm:"primaryKey;size:50" json:"name"`
	Enabled             bool      `gorm:"default:false" json:"enabled"`
	AutoOpen            bool      `gorm:"default:true" json:"auto_open"`
	AutoClose           bool      `gorm:"default:true" json:"auto_close"`
	OpenMinProfitRate   float64   `gorm:"type:decimal(10,6)" json:"open_min_profit_rate"`
	CloseMinProfitRate  float64   `gorm:"type:decimal(10,6)" json:"close_min_profit_rate"`
	MaxHoldSeconds      int64     `json:"max_hold_seconds"`
	MaxOpenPositions    int       `json:"max_open_positions"`
	CheckIntervalMillis int64     `json:"check_interval_millis"`
	UpdatedAt           time.Time `json:"updated_at"`
}

type CEXSpotPerpOpportunityLog struct {
	Key              string    `gorm:"primaryKey;size:160" json:"key"`
	OpportunityID    string    `gorm:"size:220;index" json:"opportunity_id"`
	Symbol           string    `gorm:"size:30;index" json:"symbol"`
	Direction        string    `gorm:"size:50;index" json:"direction"`
	SpotExchange     string    `gorm:"size:20;index" json:"spot_exchange"`
	PerpExchange     string    `gorm:"size:20;index" json:"perp_exchange"`
	FirstSeenAt      int64     `gorm:"index" json:"first_seen_at"`
	LastSeenAt       int64     `gorm:"index" json:"last_seen_at"`
	SeenCount        int       `json:"seen_count"`
	BestProfit       float64   `gorm:"type:decimal(20,8)" json:"best_profit"`
	BestProfitRate   float64   `gorm:"type:decimal(10,6)" json:"best_profit_rate"`
	LastProfit       float64   `gorm:"type:decimal(20,8)" json:"last_profit"`
	LastProfitRate   float64   `gorm:"type:decimal(10,6)" json:"last_profit_rate"`
	LastStatus       string    `gorm:"size:20" json:"last_status"`
	LastBlockReason  string    `gorm:"type:text" json:"last_block_reason"`
	AutoOpenedCount  int       `json:"auto_opened_count"`
	AutoRejectedNote string    `gorm:"type:text" json:"auto_rejected_note"`
	UpdatedAt        time.Time `json:"updated_at"`
}

type CEXSpotPerpAutoTrade struct {
	ID           string    `gorm:"primaryKey;size:220" json:"id"`
	PositionID   string    `gorm:"size:120;index" json:"position_id"`
	Opportunity  string    `gorm:"size:220;index" json:"opportunity"`
	Symbol       string    `gorm:"size:30;index" json:"symbol"`
	Direction    string    `gorm:"size:50;index" json:"direction"`
	SpotExchange string    `gorm:"size:20;index" json:"spot_exchange"`
	PerpExchange string    `gorm:"size:20;index" json:"perp_exchange"`
	Action       string    `gorm:"size:20;index" json:"action"`
	Reason       string    `gorm:"type:text" json:"reason"`
	Quantity     float64   `gorm:"type:decimal(20,8)" json:"quantity"`
	Notional     float64   `gorm:"type:decimal(20,8)" json:"notional"`
	Margin       float64   `gorm:"type:decimal(20,8)" json:"margin"`
	SpotValue    float64   `gorm:"type:decimal(20,8)" json:"spot_value"`
	CapitalUsed  float64   `gorm:"type:decimal(20,8)" json:"capital_used"`
	Profit       float64   `gorm:"type:decimal(20,8)" json:"profit"`
	ProfitRate   float64   `gorm:"type:decimal(10,6)" json:"profit_rate"`
	CreatedAtMs  int64     `gorm:"index" json:"created_at_ms"`
	CreatedAt    time.Time `json:"created_at"`
}

type CEXSpotPerpSimAccount struct {
	Exchange      string    `gorm:"primaryKey;size:20" json:"exchange"`
	USDT          float64   `gorm:"type:decimal(20,8)" json:"usdt"`
	PerpUSDT      float64   `gorm:"type:decimal(20,8)" json:"perp_usdt"`
	FrozenUSDT    float64   `gorm:"type:decimal(20,8)" json:"frozen_usdt"`
	SpotBalances  string    `gorm:"type:json" json:"spot_balances"`
	PerpPositions string    `gorm:"type:json" json:"perp_positions"`
	UpdatedAt     time.Time `json:"updated_at"`
}

type CEXSpotPerpSimPosition struct {
	ID              string    `gorm:"primaryKey;size:220" json:"id"`
	OpportunityID   string    `gorm:"size:220;index" json:"opportunity_id"`
	Symbol          string    `gorm:"size:30;index" json:"symbol"`
	Direction       string    `gorm:"size:50;index" json:"direction"`
	SpotExchange    string    `gorm:"size:20;index" json:"spot_exchange"`
	PerpExchange    string    `gorm:"size:20;index" json:"perp_exchange"`
	Status          string    `gorm:"size:20;index" json:"status"`
	OpenedAt        int64     `gorm:"index" json:"opened_at"`
	ClosedAt        int64     `gorm:"index" json:"closed_at"`
	Notional        float64   `gorm:"type:decimal(20,8)" json:"notional"`
	Margin          float64   `gorm:"type:decimal(20,8)" json:"margin"`
	RealizedPnL     float64   `gorm:"type:decimal(20,8)" json:"realized_pnl"`
	OpportunityJSON string    `gorm:"type:json" json:"opportunity_json"`
	TradesJSON      string    `gorm:"type:json" json:"trades_json"`
	UpdatedAt       time.Time `json:"updated_at"`
}
